addons/binding/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/SpotifyAuthService.java

```diff
@@ -0,0 +1,176 @@
+    }
+
+    /**
+     * Reads a template from file and returns the content as String.
+     *
+     * @param templateName name of the template file to read
+     * @return The content of the template file
+     * @throws IOException thrown when an HTML template could not be read
+     */
+    private String readTemplate(String templateName) throws IOException {
+        final URL index = bundleContext.getBundle().getEntry(templateName);
+
+        if (index == null) {
+            throw new FileNotFoundException(
+                    String.format("Cannot find '{}' - failed to initialize Spotify servlet", templateName));
+        } else {
+            try (InputStream inputStream = index.openStream()) {
+                return IOUtils.toString(inputStream);
+            }
+        }
+    }
+
+    /**
+     * Call with Spotify redirect uri returned State and Code values to get the refresh and access tokens and persist
+     * these values
+     *
+     * @param servletBaseURL the servlet base, which will be the Spotify redirect url
+     * @param state The Spotify returned state value
+     * @param code The Spotify returned code value
+     * @return returns the name of the Spotify user that is authorized
+     */
+    public String authorize(String servletBaseURL, String state, String code) {
+        final SpotifyAccountHandler listener = getSpotifyAuthListener(state);
+
+        if (listener == null) {
+            logger.debug(
+                    "Spotify redirected with state '{}' but no matching bridge was found. Possible bridge has been removed.",
+                    state);
+            throw new SpotifyException(ERROR_UKNOWN_BRIDGE);
+        } else {
+            return listener.authorize(servletBaseURL, code);
+        }
+    }
+
+    /**
+     * @param listener Adds the given handler
+     */
+    public void addSpotifyAccountHandler(SpotifyAccountHandler listener) {
+        if (!handlers.contains(listener)) {
+            handlers.add(listener);
+        }
+    }
+
+    /**
+     * @param handler Removes the given handler
+     */
+    public void removeSpotifyAccountHandler(SpotifyAccountHandler handler) {
+        handlers.remove(handler);
+    }
+
+    /**
+     * @return Returns all {@link SpotifyAccountHandler}s.
+     */
+    public List<SpotifyAccountHandler> getSpotifyAccountHandlers() {
+        return handlers;
+    }
+
+    /**
+     * Get the {@link SpotifyAccountHandler} that matches the given thing UID.
+     *
+     * @param thingUID UID of the thing to match the handler with
+     * @return the {@link SpotifyAccountHandler} matching the thing UID or null
+     */
+    private @Nullable SpotifyAccountHandler getSpotifyAuthListener(String thingUID) {
+        final Optional<SpotifyAccountHandler> listener = handlers.stream().filter(l -> l.equalsThingUID(thingUID))
+                .findFirst();
+        return listener.isPresent() ? listener.get() : null;
```

Comment:
I think this can be written differently `listener. orElse(null)` Consider naming optional things more explicit, what about a `maybeListener`, you will be more likely to check before calling `.get()` on it.